SCRUM-1: Create usermanager and provide user storage and retrieval functionality

ovanov opened the pull request:
This change provides commonly used functionality:
- user storage and retrieval
- connection to DynamoDB
- Dockerization
- AWS Lambda functionality

Commit:
SCRUM-1: set up-repository

Commit:
SCRUM-1: add basic functionality
- provide UserService to create user in DynamoDB
- write DynamoDB config which uses JPA to read and write
- create tables for Follows and UserInfo
- write basic integration test for Follow Repository
- dockerize application
- enable Lambda functionality for app

Commit:
SCRUM-21: dockerize usermanager

Commit:
SCRUM-20: provide create and retrieve functionality for users
- connect DynamoDB and enable functionality
- adjust UserController
- extend UserService with creation functionality
- provide DB schema to openapi Object mapping
- adjust user.yaml so that it is more concise with the defined scope

krwro commented:
You can move the docker-compose-localstack.yml into the user-manager project folder. In the yml change the context in the user-manager to context: . like this:
  user-manager:
    build:
      context: .
      dockerfile: Dockerfile
      
That way, you can easily run the docker-compose without changing folders in the project terminal.

krwro commented on src/main/java/ch/nexusnet/usermanager/service/UserService.java:
```diff
@@ -0,0 +1,56 @@
+        if (newUserInfo.isEmpty()) {
+            UserInfo createdUserInfo = saveUserToDB(newUser);
+            return mapUserInfoToUser(createdUserInfo);
+        }
```
Should the method silently return the user when that user already exists? I think raising an error telling a user already exists would be more appropriate

ovanov replied:
agreed

krwro commented on src/main/java/ch/nexusnet/usermanager/service/UserService.java:
```diff
@@ -0,0 +1,56 @@
+        if (userInfo.isPresent()) {
+            return mapUserInfoToUser(userInfo.get());
+        }
+        throw new UserNotFoundException("UserNotFound");
```
It would be helpful to add the id of the not found user in the Exception message

ovanov replied:
agreed

krwro reviewed (commented):
Looks good.

Commit:
SCRUM-21: dockerize app by using docker-compose to fully enable localstack's functionality
- minor refactorings in UserService

r-carissimi commented on src/main/java/ch/nexusnet/usermanager/service/UserService.java:
```diff
@@ -0,0 +1,63 @@
+
+    public User createUser(User newUser) throws UserAlreadyExistsException {
+        if (newUser.getId() == null) {
+            newUser.setId(UUID.randomUUID());
```
Wouldn't it be better to assign the user's UUID in the constructor?

ovanov replied:
You can, but when a user is 'created' in the front-end we need to assign a uuid in the back-end. If the front-end can provide the id via amplify, then we can refactor this

r-carissimi replied:
Didn't thought about it
Your approach is better than

r-carissimi reviewed (commented):
Looks good to me

r-carissimi committed:
SCRUM-21: lambda support on localstack

Commit:
SCRUM-1: implement new endpoint
- retrieve user by username
- enhance docker-compose-localstack.yml
- write README.md
- write tests for UserService

krwro reviewed (approved):
Looks good!
